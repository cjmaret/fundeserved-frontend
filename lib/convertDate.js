export default function convertDate(dateValue) {
  const date = new Date(dateValue.toString());
  let options = {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  };
  return new Intl.DateTimeFormat('en-US', options).format(date);
}
