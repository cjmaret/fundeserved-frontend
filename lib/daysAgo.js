export default function daysAgo(dateValue) {
  const currentDate = new Date().getDate();
  const prevDate = new Date(dateValue).getDate();
  return Math.abs(currentDate - prevDate);
}
