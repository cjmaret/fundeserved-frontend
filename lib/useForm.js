import { useEffect, useState } from 'react';

export default function useForm(initialState = {}, fileInputRef) {
  const [inputs, setInputs] = useState(initialState);
  const initialValues = Object.values(initialState).join('');

  useEffect(() => {
    setInputs(initialState);
  }, [initialValues]);

  function handleChange(e) {
    let { name, type, value } = e.target;
    console.log(e.target.files);
    if (type === 'file') {
      [value] = e.target.files;
    }
    setInputs({
      ...inputs,
      [name]: value,
    });
  }

  function resetForm() {
    setInputs(initialState);
  }

  function clearForm() {
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );
    setInputs(blankState);
  }

  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
