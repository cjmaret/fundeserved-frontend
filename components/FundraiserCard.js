import { useContext } from 'react';
import { MobileContext } from '../contexts/mobileContext';
import {
  FundraiserCardComponent,
  FundraiserParagraph,
  FundraiserImageWrapper,
  FundraiserTitle,
  FundraiserAmount,
  PercentageBarGroup,
  PercentageBarFilled,
  PercentageBar,
  FundraiserDetails,
  AmountSpan,
  CreatedOn,
  FundraiserLink,
} from './styles/styledFundraiserCard';
import { formatCentsToDollars } from '../lib/formatMoney';
import convertDate from '../lib/convertDate';

export default function FundraiserCard({ fundraiser }) {
  const mobileWidth = useContext(MobileContext);
  const fundraiserImage = fundraiser?.photo?.image?.publicUrlTransformed;
  const fundraiserDate = fundraiser?.dateCreated;

  return (
    <FundraiserCardComponent>
      <FundraiserLink href={`/fundraiser/${fundraiser.id}`} />
      <FundraiserImageWrapper>
        <img src={fundraiserImage} alt="" className="fundraiser-image" />
      </FundraiserImageWrapper>
      <FundraiserDetails>
        <FundraiserTitle>{fundraiser.name}</FundraiserTitle>
        <FundraiserParagraph>{fundraiser?.description}</FundraiserParagraph>
        <CreatedOn>Created on {convertDate(fundraiserDate)}</CreatedOn>
        <PercentageBarGroup>
          <PercentageBarFilled
            filled={(fundraiser.amount / fundraiser.goal) * 100}
          />
          <PercentageBar />
        </PercentageBarGroup>
        <FundraiserAmount>
          <AmountSpan>
            {formatCentsToDollars(fundraiser?.amount)} raised
          </AmountSpan>
          of {formatCentsToDollars(fundraiser.goal)}
        </FundraiserAmount>
      </FundraiserDetails>
    </FundraiserCardComponent>
  );
}
