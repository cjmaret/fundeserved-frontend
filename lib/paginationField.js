import { PAGINATION_COUNT_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // allows custom
    read(existing = [], { args, cache }) {
      const { skip, first } = args;

      // reads number of items on the page from the cache
      const data = cache.readQuery({ query: PAGINATION_COUNT_QUERY });
      const count = data?._allFundraisersMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      // check for existing items
      const items = existing.slice(skip, skip + first).filter((x) => x);
      // If there are items, and there aren't enough items to satisfy how many were requested, and we're on the last page, then just send the data
      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        return false; // no items, need to fetch from network
      }
      if (items.length) {
        console.log(
          `There are ${items.length} items in the cache, sending to Apollo`
        );
        return items;
      }
    },
    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      // merge runs when Apollo client comes back from the network with items
      console.log(`Merging items from the network ${incoming.length}`);
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; i++) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}
