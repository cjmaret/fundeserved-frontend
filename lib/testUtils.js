import casual from 'casual';
import { PAGINATION_COUNT_QUERY } from '../components/Pagination';

casual.seed(777);

const fakeFundraiser = () => ({
  id: 'abc123',
  name: 'test-fundraiser',
  description: 'fundraiser description',
  status: 'ACTIVE',
  amount: 500,
  goal: 600,
  dateCreated: '2022-08-25T16:00:00.000Z',
  user: {
    id: '123',
  },
  photo: {
    id: 'abc123',
    altText: 'test-fundraiser',
    image: {
      publicUrlTransformed: 'dog.jpg',
    },
  },
  donations: [
    {
      total: 50,
      dateCreated: '2022-09-25T16:00:00.000Z',
      user: {
        name: 'Colin',
        avatar: {
          publicUrlTransformed: 'dog.jpg',
        },
      },
    },
  ],
});

const fakeUser = (overrides) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  createdOn: '2022-05-25T16:00:00.000Z',
  avatar: {
    publicUrlTransformed: 'colin.jpg',
  },
  fundraisers: [
    {
      id: '123',
      name: 'fundraiser1',
      description: 'fundraiser',
      amount: 5000,
      goal: 6000,
      dateCreated: '2022-09-25T16:00:00.000Z',
      photo: {
        image: {
          publicUrlTransformed: 'fundraiser1.jpg',
        },
      },
    },
  ],
  donations: [
    {
      id: '123',
      total: 25,
      dateCreated: '2022-05-26T16:00:00.000Z',
      fundraiser: {
        id: '456',
        name: 'fundraiser2',
        description: 'fundraiser',
        photo: {
          image: {
            publicUrlTransformed: 'fundraiser2.jpg',
          },
        },
      },
    },
  ],
  permissions: ['OVERLORD'],

  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
});


class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_COUNT_QUERY },
      result: {
        data: {
          _allFundraisersMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeFundraiser,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
