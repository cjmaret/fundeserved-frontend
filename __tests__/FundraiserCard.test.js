import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import FundraiserCard from '../components/FundraiserCard';
import { fakeFundraiser } from '../lib/testUtils';

const fundraiser = fakeFundraiser();

describe('<FundraiserCard/>', () => {
  it('renders out price tag, link, and title', () => {
    const { container, debug } = render(
      <MockedProvider>
        <FundraiserCard fundraiser={fundraiser} />
      </MockedProvider>
    );
    const amountRaised = screen.getByText('$5 raised');
    expect(amountRaised).toBeInTheDocument();
    const link = container.querySelector('a');
    expect(link).toHaveAttribute('href', '/fundraiser/abc123');
    const title = container.querySelector('h1');
    expect(title).toHaveTextContent('test-fundraiser');
  });

  it('renders and matches the snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <FundraiserCard fundraiser={fundraiser} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders the image properly', () => {
    const { container, debug } = render(
      <MockedProvider>
        <FundraiserCard fundraiser={fundraiser} />
      </MockedProvider>
    );
    const img = screen.getByAltText(fundraiser.name);
    expect(img).toBeInTheDocument();
  });
});
