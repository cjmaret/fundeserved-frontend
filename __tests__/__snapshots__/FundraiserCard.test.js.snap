// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FundraiserCard/> renders and matches the snapshot 1`] = `
<div>
  <div
    class="styledFundraiserCard__FundraiserCardComponent-sc-2xi0y7-0 fTrHup"
  >
    <a
      class="styledFundraiserCard__FundraiserLink-sc-2xi0y7-1 dgaaRK"
      href="/fundraiser/abc123"
    />
    <div
      class="styledFundraiserCard__FundraiserImageWrapper-sc-2xi0y7-2 osxpu"
    >
      <img
        alt="test-fundraiser"
        class="fundraiser-image"
        src="dog.jpg"
      />
    </div>
    <div
      class="styledFundraiserCard__FundraiserDetails-sc-2xi0y7-3 kYeYAz"
    >
      <h1
        class="styledFundraiserCard__FundraiserTitle-sc-2xi0y7-4 QTDWe"
      >
        test-fundraiser
      </h1>
      <p
        class="styledFundraiserCard__FundraiserParagraph-sc-2xi0y7-5 eEGTVv"
      >
        fundraiser description
      </p>
      <div
        class="styledFundraiserCard__CreatedOn-sc-2xi0y7-6 gbDDxP"
      >
        Created on 
        August 25, 2022
      </div>
      <div
        class="styledFundraiserCard__PercentageBarGroup-sc-2xi0y7-7 fyhWNB"
      >
        <div
          class="styledFundraiserCard__PercentageBarFilled-sc-2xi0y7-9 dIZjJJ"
        />
        <div
          class="styledFundraiserCard__PercentageBar-sc-2xi0y7-8 bZwITN"
        />
      </div>
      <p
        class="styledFundraiserCard__FundraiserAmount-sc-2xi0y7-10 eAXCly"
      >
        <span
          class="styledFundraiserCard__AmountSpan-sc-2xi0y7-11 dDlhku"
        >
          $5
           raised
           
        </span>
        of 
        $6
      </p>
    </div>
  </div>
</div>
`;
